src/Core/Checkout/Cart/CartRuleLoader.php

```diff
@@ -220,12 +220,25 @@ private function detectTaxType(SalesChannelContext $context, float $cartNetAmoun
         }
 
         $country = $context->getShippingLocation()->getCountry();
+        $customer = $context->getCustomer();
 
-        $isReachedCustomerTaxFreeAmount = $country->getCustomerTax()->getEnabled() && $this->isReachedCountryTaxFreeAmount($context, $country, $cartNetAmount);
-        $isReachedCompanyTaxFreeAmount = $this->taxDetector->isCompanyTaxFree($context, $country) && $this->isReachedCountryTaxFreeAmount($context, $country, $cartNetAmount, CountryDefinition::TYPE_COMPANY_TAX_FREE);
+        // Check if customer is a B2B customer (has company)
+        $isB2BCustomer = $customer && !empty($customer->getCompany());
```

Comment:
Note there is also the `accountType`: https://github.com/shopware/shopware/blob/trunk/src/Core/Checkout/Customer/CustomerDefinition.php#L158